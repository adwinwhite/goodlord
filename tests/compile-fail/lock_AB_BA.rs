use deadlock_proof::LockAfter;
use deadlock_proof::impl_lock_order;
use deadlock_proof::LockToken;
use deadlock_proof::GoodLock;
use deadlock_proof::Unlocked;


struct A;
struct B;

impl_lock_order!(Unlocked => A);
impl_lock_order!(A => B);

fn main() {
    let mut unlocked_token = LockToken::UNLOCKED;
    let lock_a: GoodLock<A, _> = GoodLock::new(0_i32);
    let lock_b: GoodLock<B, _> = GoodLock::new(0_i32);
    let (b, locked_b) = lock_b.lock(&mut unlocked_token);
    let (a, _) = lock_a.lock(&mut locked_b);
    println!("{}, {}", a, b);
}
