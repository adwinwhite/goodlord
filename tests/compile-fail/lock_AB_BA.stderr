error[E0308]: mismatched types
  --> tests/compile-fail/lock_AB_BA.rs:19:30
   |
19 |     let (a, _) = lock_a.lock(&mut locked_b);
   |                         ---- ^^^^^^^^^^^^^ expected `&mut LockToken<Unlocked>`, found `&mut LockToken<B>`
   |                         |
   |                         arguments to this method are incorrect
   |
   = note: expected mutable reference `&mut LockToken<Unlocked>`
              found mutable reference `&mut LockToken<B>`
note: method defined here
  --> src/lib.rs
   |
   |     pub fn lock<'t, 's, PrevId>(
   |            ^^^^
